### PR TITLE
[WIP] server side http telemetry

### PR DESCRIPTION
**Pull request by jennyf19**
@bgavrilMS

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Internal/Requests/DeviceCodeRequest.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@
 using Microsoft.Identity.Client.ApiConfig.Parameters;
 using Microsoft.Identity.Client.Core;
 using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.TelemetryCore;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/Requests/DeviceCodeRequest.cs]: not needed?

File: src/client/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs [hunk 2]
```diff
@@ -235,6 +236,9 @@ private Uri CreateAuthorizationUri(bool addPkceAndState = false)
                 authorizationRequestParameters[OAuth2Parameter.Prompt] = _interactiveParameters.Prompt.PromptValue;
             }
 
+            authorizationRequestParameters[TelemetryConstants.XClientLastRequest] = ServiceBundle.TelemetryManager.FetchAndResetPreviousHttpTelemetryContent();
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs]: I though we are sending this as headers? Here you are sending this as form params. Also, is it confirmed that the server accepts telemetry on the /authorize endpoint?

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryConstants.cs [hunk 3]
```diff
@@ -16,4 +16,11 @@ internal static class TelemetryError
     //    public const string MsalCacheEventCount = "msal.cache_event_count";
     //    public const string MsalUiEventCount = "msal.ui_event_count";
     //}
+
+    internal static class TelemetryConstants
+    {
+        public const string HttpTelemetrySchemaVersion1 = "1|";
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryConstants.cs]: Please set the pipe as a different constant.

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs [hunk 4]
```diff
@@ -64,6 +72,8 @@ public void StartEvent(EventBase eventToStart)
 
         public void StopEvent(EventBase eventToStop)
         {
+            ProcessEventsForPreviousHttpTelemetryContent(eventToStop); //try movnig up to start
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs]: Please consolidate - set it on start or on stop.

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs [hunk 5]
```diff
@@ -137,6 +147,93 @@ public void Flush(string correlationId)
             Callback?.Invoke(eventsToFlush.Cast<Dictionary<string, string>>().ToList());
         }
 
+        public string FetchAndResetPreviousHttpTelemetryContent()
+        {
+            if (!string.IsNullOrEmpty(_previousTelemetryPayload.ApiId))
+            {
+                string csv = CreateCSVForPreviousHttpTelem();
+                _currentTelemetryPayload.ResetLastErrorCode();
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs]: Should be `_previousTelemetryPayload `

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs [hunk 6]
```diff
@@ -137,6 +147,93 @@ public void Flush(string correlationId)
...
+
+            string csvString = string.Join(",", myValues);
+            csvString = TelemetryConstants.HttpTelemetrySchemaVersion1 + csvString;
+            return csvString;
+        }
+
+        private string CreateCSVForCurrentHttpTelem()
+        {
+            // csv expected format:
+            // 1|api_id|platform_config
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs]: The actual fomat seems to be 1|api_id|correlation_id

File: src/client/Microsoft.Identity.Client/WsTrust/WsTrustWebRequestManager.cs [hunk 7]
```diff
@@ -26,8 +28,10 @@ public WsTrustWebRequestManager(IHttpManager httpManager)
         /// <inheritdoc/>
         public async Task<MexDocument> GetMexDocumentAsync(string federationMetadataUrl, RequestContext requestContext)
         {
+            IDictionary<string, string> headers = CreateHttpTelemetryHeaders(requestContext);
+
             var uri = new UriBuilder(federationMetadataUrl);
-            HttpResponse httpResponse = await _httpManager.SendGetAsync(uri.Uri, null, requestContext.Logger).ConfigureAwait(false);
+            HttpResponse httpResponse = await _httpManager.SendGetAsync(uri.Uri, headers, requestContext.Logger).ConfigureAwait(false);
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/WsTrust/WsTrustWebRequestManager.cs]: I don't think the Ws Trust endpoint supports this, nor do I think they will ever support this, so might as well not send the data.

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/WsTrust/WsTrustWebRequestManager.cs]: That said, we should definitely send the MsalIDParameters here.

File: src/client/Microsoft.Identity.Client/WsTrust/WsTrustWebRequestManager.cs [hunk 8]
```diff
@@ -109,16 +113,37 @@ public async Task<MexDocument> GetMexDocumentAsync(string federationMetadataUrl,
...
 
             return httpResponse.StatusCode == System.Net.HttpStatusCode.OK
                 ? JsonHelper.DeserializeFromJson<UserRealmDiscoveryResponse>(httpResponse.Body)
                 : null;
         }
+
+        private IDictionary<string, string> CreateHttpTelemetryHeaders(RequestContext requestContext)
+        {
+            IDictionary<string, string> httpTelemetryHeader = new Dictionary<string, string>();
+            httpTelemetryHeader[TelemetryConstants.XClientLastRequest] = requestContext.ServiceBundle.TelemetryManager.FetchAndResetPreviousHttpTelemetryContent();
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/WsTrust/WsTrustWebRequestManager.cs]: An authentication flow involves talking to the several endpoints - discovery, oidc, ws-trust, authroize, token. However, I do not think the server side is ready to accept data on all of these - my understanding is that it only accepts the data on /token (and maybe /authorize?). 

And since you're resetting the last error code everytime you send the telemetry data, we'll most likely lose it. So I propose either: 

1. Only send telemetry on /token endpoint OR
2. Only reset the last error code when sending to /token endpoint

File: tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHttpMessageHandler.cs [hunk 9]
```diff
@@ -96,6 +98,13 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
                 }
             }
 
+            if (ExpectedHeaders != null)
+            {
+                Assert.IsTrue(ActualRequestMessage.Headers.Contains(TelemetryConstants.XClientLastRequest));
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHttpMessageHandler.cs]: This assertion is not strong enough. You can expose `ActualHeaders` and then validate that you're sending the correct telemetry in a variatey of use cases.

File: tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHttpMessageHandler.cs [hunk 10]
```diff
@@ -96,6 +98,13 @@ protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage reques
...
+                Assert.IsTrue(ActualRequestMessage.Headers.Contains(TelemetryConstants.XClientCurrentTelemetry));
+                ExpectedHeaders = null;
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHttpMessageHandler.cs]: Why is this needed?

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs [hunk 11]
```diff
@@ -137,6 +147,93 @@ public void Flush(string correlationId)
...
+        private void ProcessEventsForPreviousHttpTelemetryContent(EventBase events)
+        {
+            foreach (var ev in events)
+            {
+                if (string.Equals(ev.Key, MsalTelemetryBlobEventNames.ApiIdConstStrKey))
+                {
+                    _previousTelemetryPayload.ApiId = ev.Value;
+                }
+
+                if (string.Equals(ev.Key, MsalTelemetryBlobEventNames.ApiErrorCodeConstStrKey))
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs]: I don't understand where this data is written since you've just defined this key, but you are not using it anywhere.

File: tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHttpMessageHandler.cs (hunk 9 above)
Comment (jennyf19) [reply tests/Microsoft.Identity.Test.Common/Core/Mocks/MockHttpMessageHandler.cs]: i added more tests

---
In reply to: [325641770](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1389#discussion_r325641770) [](ancestors = 325641770)

File: src/client/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs (hunk 2 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs]: as we discussed...only sending to /token endpoint now

---
In reply to: [325635909](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1389#discussion_r325635909) [](ancestors = 325635909)

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs (hunk 6 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs]: the format was right...i removed correlation_id

---
In reply to: [325638151](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1389#discussion_r325638151) [](ancestors = 325638151)

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs (hunk 11 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs]: not sure i follow.

---
In reply to: [325790729](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1389#discussion_r325790729) [](ancestors = 325790729)

File: src/client/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs (hunk 2 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs]: image

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs [hunk 12]
```diff
@@ -137,6 +146,93 @@ public void Flush(string correlationId)
             Callback?.Invoke(eventsToFlush.Cast<Dictionary<string, string>>().ToList());
         }
 
+        public string FetchAndResetPreviousHttpTelemetryContent()
+        {
+            if (!string.IsNullOrEmpty(_previousTelemetryPayload.ApiId))
+            {
+                string csv = CreateCSVForPreviousHttpTelem();
+               // _previousTelemetryPayload.ResetLastErrorCode();
```
Comment (MarkZuber) [inline src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs]: dead code?

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs [hunk 13]
```diff
@@ -137,6 +146,93 @@ public void Flush(string correlationId)
...
+                return string.Empty;
+            }
+        }
+
+        public string FetchAndResetCurrentHttpTelemetryContent()
+        {
+            if (!string.IsNullOrEmpty(_currentTelemetryPayload.ApiId))
+            {
+                string csv = CreateCSVForCurrentHttpTelem();
+               //_currentTelemetryPayload.ResetLastErrorCode();
```
Comment (MarkZuber) [inline src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs]: dead code?

File: src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs [hunk 14]
```diff
@@ -64,6 +70,9 @@ public void StartEvent(EventBase eventToStart)
 
         public void StopEvent(EventBase eventToStop)
         {
+            ProcessEventsForCurrentHttpTelemetryContent(eventToStop);
```
Comment (MarkZuber) [inline src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryManager.cs]: Does this need locking?  We could have more than one request going on at the same time which means we could be calling StopEvent in the telemetry manager more than once.  and this call pattern isn't idempotent or atomic.

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs [hunk 15]
```diff
@@ -25,7 +26,18 @@ internal class AcquireTokenCommonParameters
 
         public void AddApiTelemetryFeature(ApiTelemetryFeature feature)
         {
-            _apiTelemetry[MatsConverter.AsString(feature)] = "true";
+            _apiTelemetry[MatsConverter.AsString(feature)] = TelemetryConstants.True;
+        }
+
+        public void AddApiTelemetryFeature(ApiTelemetryFeature feature, bool isTrue)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenCommonParameters.cs]: Is this feature (tracking the value of some flags) needed? It seems like a scope creep. Please remove if.

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemetryContent.cs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+                evt.TryGetValue(MsalTelemetryBlobEventNames.ForceRefreshId, out string forceRefresh);
+
+                ApiId = apiId ?? string.Empty;
+                CorrelationId = correlationId ?? string.Empty;
+                LastErrorCode = errorCode ?? string.Empty;
+                ForceRefresh = forceRefresh ?? string.Empty;
+            }
+        }
+
+        public string LastErrorCode { get; set; } = string.Empty;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemetryContent.cs]: Do you need setters and default values?

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemetryContent.cs [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+        public string GetCsvAsPrevious()
+        {
+            if (string.IsNullOrWhiteSpace(ApiId))
+            {
+                return string.Empty;
+            }
+
+            // csv expected format:
+            // 1|api_id,correlation_id,last_error_code|platform_config
+            string[] myValues = new string[] {
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemetryContent.cs]: try to refer to local variables without the pronoun `my`. Instead, give a name that has meaning in the local context, for example `dataToCollect`

File: src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemetryContent.cs [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+            string csvString = string.Join(",", myValues);
+            csvString = $"{TelemetryConstants.HttpTelemetrySchemaVersion1}{TelemetryConstants.HttpTelemetryPipe}{csvString}{TelemetryConstants.HttpTelemetryPipe}";
+            return csvString;
+        }
+
+        public string GetCsvAsCurrent()
+        {
+            // csv expected format:
+            // 1|api_id,force_refresh|platform_config
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/TelemetryCore/Internal/HttpTelemetryContent.cs]: Hmm, there is a mismatch in semantics between MATS and HTTP telemetry. In fact, there are 3 values for all flags similar to `forceRefresh`

- user doesn't care and did not set it
- user set it to true
- user set it to false

MATS captures this 3-state because it will not log any event when user doesn't set a value explicitly. HTTP merges the meaning of "user did not set it" and "user set it to false".
